src/frontmatter/frontmatter.ts

```diff
@@ -0,0 +1,76 @@
+import * as yaml from 'js-yaml';
+import * as Unist from 'unist';
+import { parseWithPointers } from '../parseWithPointers';
+import { stringify } from '../stringify';
+import { MarkdownParserResult } from '../types';
+import { IFrontmatter } from './types';
+import { countNewLines, shiftLines } from './utils';
+
+export class Frontmatter<T extends object = any> implements IFrontmatter<T> {
+  private readonly document: MarkdownParserResult;
+  private readonly node: Unist.Literal | null;
+  private readonly root: Unist.Parent;
+  private readonly properties: T | null;
+
+  constructor(data: MarkdownParserResult | string) {
+    this.document = typeof data === 'string' ? parseWithPointers(data) : data;
+
+    const root = this.document.ast;
+    if (root.type !== 'root') {
+      throw new TypeError('Malformed yaml was provided');
+    }
+
+    this.root = root;
+    this.node =
+      root.children.length > 0 && root.children[0].type === 'yaml' ? (root.children[0] as Unist.Literal) : null;
+
+    this.properties = this.node !== null ? yaml.safeLoad(String(this.node.value)) : null;
+  }
+
+  private updateDocument() {
+    const oldValue = this.node!.value;
+
+    this.node!.value = yaml
+      .safeDump(this.properties, {
+        flowLevel: 1,
+        indent: 2,
+      })
+      .trim();
+
+    const diff = countNewLines(this.node!.value as string) - countNewLines(oldValue as string);
+
+    if (diff !== 0) {
+      shiftLines(this.root, diff);
+    }
+  }
+
+  public getAll(): T | void {
+    if (this.properties !== null) {
+      return this.properties;
+    }
+  }
+
+  public get(prop: keyof T): T[keyof T] | void {
+    if (this.properties !== null) {
+      return this.properties[prop];
+    }
+  }
+
+  public set(prop: keyof T, value: T[keyof T]) {
+    if (this.properties !== null) {
+      this.properties[prop] = value;
```

Comment:
Can we use lodash [get](https://www.npmjs.com/package/lodash.get), [set](https://www.npmjs.com/package/lodash.set), and [unset](https://www.npmjs.com/package/lodash.unset) here instead?

For example on the homepage/blog, we have some pretty deeply nested frontmatter objects 😬 

https://raw.githubusercontent.com/stoplightio/stoplight.io/develop/netlify/blog-posts/10-000-apis-under-the-sea.md

Reply:
@lottamus sure!
I was unaware of the case you mentioned, so did in a more traditional fashion.

Reply:
no worries! i think we will be able to use this for the homepage too when we have our own admin panel so might make sense to add it. those 3 lodash functions should be a small dep, right?

Reply:
Yup, it shouldn't add any significant overhead. I believe Lodash is likely to be already included anyway by one of our dependencies.